app/public/static/js/components/ordered-puzzle.js

```diff
@@ -0,0 +1,196 @@
+      rawImageEl.setAttribute("index", i);
+
+      rawImageEl.onload = function (event) {
+        const index = event.target.getAttribute("index");
+        const puzzlePiece = puzzlePieceCache[index];
+
+        // Calculate the puzzle piece width and height based on image ratio
+        const ratio = rawImageEl.width / rawImageEl.height;
+        const scaleBy = data.hasOwnProperty("scaleBy") ? data.scaleBy : 3;
+        let puzzlePieceWidth = scaleBy * ratio;
+        let puzzlePieceHeight = scaleBy;
+        puzzlePiece.setAttribute("width", puzzlePieceWidth);
+        puzzlePiece.setAttribute("height", puzzlePieceHeight);
+        puzzlePiece.setAttribute("class", "draggable link");
+
+        if (data.randomizePos) {
+          // Randomizes the position of the puzzle piece on the blackboard
+          puzzlePiece.setAttribute("position", {
+            x:
+              (Math.random() - 0.5) *
+              (blackboard.getAttribute("geometry").width - puzzlePieceWidth),
+            y:
+              (Math.random() - 0.5) *
+              (blackboard.getAttribute("geometry").height - puzzlePieceHeight),
+            z: 0,
+          });
+        } else {
+          puzzlePiece.setAttribute("position", {
+            x: data.images[index].x * puzzlePieceWidth,
+            y: data.images[index].y * puzzlePieceHeight,
+            z: 0,
+          });
+        }
+
+        if (data.useTargets) {
+          // Create target
+          let textBoxProp = JSON.parse(JSON.stringify(data.images[index]));
+          textBoxProp.color = "yellow";
+          textBoxProp.width = puzzlePieceWidth;
+          textBoxProp.height = puzzlePieceHeight;
+          textBoxProp.xTarget = textBoxProp.xTarget * puzzlePieceWidth;
+          textBoxProp.yTarget = textBoxProp.yTarget * puzzlePieceHeight;
+
+          let target = document.createElement("a-entity");
+          target.setAttribute("id", "puzzle-target-" + index);
+          target.setAttribute(
+            "target-box",
+            "jsonData",
+            JSON.stringify(textBoxProp)
+          );
+          el.appendChild(target);
+
+          // Set image's target location
+          puzzlePiece.setAttribute(
+            "xTarget",
+            data.images[index].xTarget * puzzlePieceWidth
+          );
+          puzzlePiece.setAttribute(
+            "yTarget",
+            data.images[index].yTarget * puzzlePieceHeight
+          );
+          puzzlePiece.setAttribute("onTarget", false);
+
+          puzzlePiece.addEventListener("dragend", function (event) {
+            event.target.setAttribute("onTarget", isOnTarget(event.target));
+            if (isPuzzleComplete(puzzlePieceCache)) {
+              // emit event to update state
+              el.sceneEl.emit("solvedObject", { id: data.id });
+              // emit event to close popup (after 1 second for now) - TODO: disabled for now
+              // need designer input on whether we should autoclose this
```

Comment:
Discussed offline, but my thought is:
- should not autoclose as these puzzles might need some time to derive the meaning / clue from
- even if it's the last puzzle, we should not move them into the next room until the popup is closed